cmd/shim/secret_scrub.go

```diff
@@ -6,80 +6,51 @@ import (
-// It is safe to write to it concurrently.
-type SecretScrubWriter struct {
-	mu sync.Mutex
-	w  io.Writer
+var (
+	// scrubString will be used as replacement for found secrets:
+	scrubString = []byte("***")
+)
 
-	secretValues []string
+type SecretScrubReader struct {
+	io.Reader
 }
 
-// Write scrubs secret values from b and replace them with `***`.
-func (w *SecretScrubWriter) Write(b []byte) (int, error) {
-	w.mu.Lock()
-	defer w.mu.Unlock()
-
-	scrubbedBytes := scrubSecretBytes(w.secretValues, b)
+func NewSecretScrubReader(r io.Reader, currentDirPath string, fsys fs.FS, env []string, secretsToScrub core.SecretToScrubInfo) (io.Reader, error) {
+	secrets := loadSecretsToScrubFromEnv(env, secretsToScrub.Envs)
 
-	_, err := w.w.Write(scrubbedBytes)
+	fileSecrets, err := loadSecretsToScrubFromFiles(currentDirPath, fsys, secretsToScrub.Files)
 	if err != nil {
-		return -1, err
+		return nil, fmt.Errorf("could not load secrets from file: %w", err)
 	}
+	secrets = append(secrets, fileSecrets...)
 
-	return len(b), err
-}
-
-func scrubSecretBytes(secretValues []string, b []byte) []byte {
-	s := string(b)
-	ss := strings.Split(s, "\n")
-
-	out := make([]string, 0, len(ss))
-
-	for _, line := range ss {
-		for _, secretValue := range secretValues {
-			var secretLines []string
-
-			lines := strings.Split(secretValue, "\n")
-			secretLines = append(secretLines, lines...)
-			for _, secretLine := range secretLines {
-				secretLine := strings.TrimSpace(secretLine)
-				if secretLine == "" {
-					continue
-				}
-				// FIXME: I think we can do better
-				line = strings.ReplaceAll(line, secretLine, "***")
-			}
+	secretAsBytes := make([][]byte, 0)
+	for _, v := range secrets {
+		// Skip empty env:
+		if len(v) == 0 {
+			continue
 		}
-		out = append(out, line)
+		secretAsBytes = append(secretAsBytes, []byte(v))
 	}
 
-	s = strings.Join(out, "\n")
-
-	return []byte(s)
-}
-
-// NewSecretScrubWriter replaces known secrets by "***".
-// The value of the secrets referenced in secretsToScrub are loaded either
-// from env or from the fs accessed at currentDirPath.
-func NewSecretScrubWriter(w io.Writer, currentDirPath string, fsys fs.FS, env []string, secretsToScrub core.SecretToScrubInfo) (io.Writer, error) {
-	secrets := loadSecretsToScrubFromEnv(env, secretsToScrub.Envs)
-
-	fileSecrets, err := loadSecretsToScrubFromFiles(currentDirPath, fsys, secretsToScrub.Files)
-	if err != nil {
-		return nil, fmt.Errorf("could not load secrets from file: %w", err)
+	replaceChain := make([]transform.Transformer, 0)
+	for _, s := range secretAsBytes {
+		replaceChain = append(
+			replaceChain,
+			replace.Bytes(s, scrubString),
+		)
+	}
+	secretScrubReader := &SecretScrubReader{
+		Reader: replace.Chain(r, replaceChain...),
 	}
-	secrets = append(secrets, fileSecrets...)
 
-	return &SecretScrubWriter{
-		w:            w,
-		secretValues: secrets,
-	}, nil
+	return secretScrubReader, nil
```

Comment:
Seems like we don't need this wrapping type?

```suggestion
	return replace.Chain(r, replaceChain...), nil
```

Reply:
Agree, just did the change